# Scenario

Let's say that you're working for a fictional productivity software company that is looking for ways to increase the number of people who pay for their software. The way that the software is currently set up, users can download and use the software free of charge, for a 7-day trial. After the end of the trial, users are required to pay for a license to continue using the software.

One idea that the company wants to try is to change the layout of the homepage to emphasize more prominently and higher up on the page that there is a 7-day trial available for the company's software. The current fear is that some potential users are missing out on using the software because of a lack of awareness of the trial period. If more people download the software and use it in the trial period, the hope is that this entices more people to make a purchase after seeing what the software can do.

<details>
  <summary>Click to see more!</summary>
  
    In this case study, you'll go through steps for planning out an experiment to test the new homepage. You will start by constructing a user funnel and deciding on metrics to track. You'll also perform experiment sizing to see how long it should be run. Afterwards, you'll be given some data collected for the experiment, perform statistical tests to analyze the results, and come to conclusions regarding how effective the new homepage changes were for bringing in more users.
</details>

# Building a Funnel

### Goal of the study
Redesign the homepage to increase the number of people who download the software and eventually the number of people who purchase a license.

The company's website has five main sections:

1. the homepage;
2. a section with additional information, gallery, and examples;
3. a page for users to download the software;
4. a page for users to purchase a license; and
5. a support sub-site with documentation and FAQs for the software.


<details>
    <summary>Click to see more!</summary>
    
    For the software itself, the website requires that users create an account in order to download the software program. The program is usable freely for seven days after download. When the trial period is hit, the program will bring up a dialog box that takes the user to the license page. After purchasing a license, the user will receive a unique code associated with their site account. This code can then be used with the program to register it with that user, and the program can be used thereafter without issue.
</details>

### Question : Expected flow
What steps do you expect typical visitors to take from their initial visit to the webpage through purchasing a license for continued use of the program? Are there any 'typical' steps that certain visitors might not take?

In [1]:
# answer here

<details>
  <summary>Example answer here</summary>
  
A straightforward flow might include the following steps:

- Visit homepage
- Visit download page
- Sign up for an account
- Download software
- After 7-day trial, software takes user to license-purchase page
- Purchase license

    Note that it is possible for the visitor to drop from the flow after each step, forming a funnel. There might be additional steps that a user might take between visiting the homepage and visiting the download page that aren't accounted for in the above flow. For example, someone might want to check out the additional informational pages before visiting the download page, or even visit the license purchase page to check the license price before even deciding to download. Considering the amount of browsing that a visitor could perform on the page, it might be simplest just to track whether or not a user gets to the download page at some point, without worrying about the many paths that they could have taken to get there.
</details>

### Question: Atypical events
Consider the webpage as a whole. What kinds of events might occur outside of the expected flow for the experiment that might interfere with measuring the effects of our manipulation?

In [2]:
# answer here

<details>
  <summary>Example answer here</summary>
   
  There are a few events in the expected flow that might not correspond with the visitors we want to target. For example, there might be users on the homepage who aren't new users. Users who already have a license might just be visiting the homepage as a way to access the support sub-site. A user who wants to buy a license might also come in to the license page through the homepage, rather than directly from the software.

When it comes to license purchasing, it's possible that users don't come back after exactly seven days. Some users might come back early and make their purchase during their trial period. Alternatively, a user might end up taking more than seven days to decide to make their purchase, coming back days after the end of the trial. Anticipating scenarios like this can be useful for planning the design, and coming up with metrics that come as close as possible to measuring desired effects.

</details>

# Deciding on metrics

From our user funnel, we should consider two things: where and how we should split users into experiment groups, and what metrics we will use to track the success or failure of the experimental manipulation. The choice of unit of diversion (the point at which we divide observations into groups) may affect what metrics we can use, and whether the metrics we record should be considered invariant or evaluation metrics. To start, decide on a unit of diversion and brainstorm some ideas for metrics to capture.

<details>
    <summary>Click to expand</summary>

To be clear, the overall plan is to test the effect of the new homepage using a true experiment; in particular, we'll be using an A/B testing framework. This means that prospective users should be split into two groups. The control, or 'A' group, will see the old homepage, while the experimental, or 'B' group, will see the new homepage that emphasizes the 7-day trial. </details>

### Question 
By which means should we divide visitors into our experimental and control groups?

<details>
    <summary>Example answer here</summary>

A cookie-based diversion seems best in this case for dividing visitors into experimental groups since we can split visitors on their initial visit and it's fairly reliable for tracking. We can assign a cookie to each visitor upon their first page hit, which allows them to be separated into the control and experimental groups. Cookies also allow tracking of each visitor hitting each page, recording whether or not they eventually hit the download page and then whether or not they actually register an account and perform the download. 
    
The usual cookie-based diversion issues apply: we can get some inconsistency in counts if users enter the site via incognito window, different browsers, or cookies that expire or get deleted before they make a download. This kind of assignment 'dilution' could dampen the true effect of our experimental manipulation. As a simplification, however, we'll assume that this kind of assignment dilution will be small, and ignore its potential effects.
    
Other methods: event-based diversion (i.e. pageview), account-based divsersion (i.e. User ID).
    
An event-based diversion (like a pageview) can provide many observations to draw conclusions from, but doesn't quite hit the mark for this case. If the condition changes on each pageview, then a visitor might get a different experience on each homepage visit. Event-based diversion is much better when the changes aren't as easily visible to users, to avoid disruption of experience. In addition, pageview-based diversion would let us know how many times the download page was accessed from each condition, but can't go any further in tracking how many actual downloads were generated from each condition.
    
Diverting based on account or user ID can be stable, but it's not the right choice in this case. Since visitors only register after getting to the download page, this is too late to introduce the new homepage to people who should be assigned to the experimental condition.    
</details>

### Question
Write down some potential metrics or ways of objectively measuring things related to evaluating the success of the experiment. You don't need to decide on invariant or evaluation metrics here: that'll be discussed in the following section.

<details>
    <summary>Example answer here</summary>
    
In terms of metrics, we might want to keep track of the number of cookies that are recorded in different parts of the website. In particular, the number of cookies on the homepage, download page, and account registration page (in order to actually make the download) could prove useful. We can track the number of licenses purchased through the user accounts, each of which can be linked back to a particular condition. Though it hasn't been specified, it's also possible that the software includes usage statistics that we could track.

The above metrics are all based on absolute counts. We could instead perform our analysis on ratios of those counts. For example, we could be interested in the proportion of downloads out of all homepage visits. License purchases could be stated as a ratio against the number of registered users (downloads) or the original number of cookies.
</details>

### Question
Below, you will decide for each of the proposed metrics whether or not you would want to use them as an invariant metric or an evaluation metric.

An invariant metric is an objective measure that you should expect will not vary between conditions and that indicate equivalence between groups. Evaluation metrics, on the other hand, represent measures where you expect there will be differences between groups, and whose differences should say something meaningful about your experimental manipulation.

1. Number of cookies @ homepage
2. Number of cookies @ download page
3. Number of user ids / downloads
4. Number of license purchases
5. Mean software usage time during trial
6. Ratio: # downloads / # cookies
7. Ratio: # licenses / # cookies
8. Ratio: # licenses / # user IDs

<details>
    <summary>Example answer here</summary>
    
`Invariant:` 1, each visitor should have an equal chance of seeing each homepage. Since visitors come in without any additional information (e.g. account info) and the change effected by the experimental manipulation comes in right at the start, there aren't other invariant metrics we should worry about.
    
`Evaluation:` 4, 6, 7
    
Count-based metrics at other parts of the process seem like natural choices: the number of times the software was downloaded and the number of licenses purchased are exactly what we want to change with the new homepage. The issue is that even though we expect the number of cookies assigned to each group to be about the same, it's much more likely than not they they won't be exactly the same. Instead, we should prefer using the download rate (# downloads / # cookies) and purchase rate (# licenses / # cookies) relative to the number of cookies as evaluation metrics. Using these ratios allows us to account for slight imbalances between groups.    
    
As for the other proposed metrics, the ratio between the number of licenses and number of downloads is potentially interesting, but not as direct as the other two ratios discussed above. It's possible that the manipulation increases both the number of downloads and number of licenses, but increases the former to a much higher rate. In this case, the licenses-to-downloads ratio might be worse off for the new homepage compared to the old, even though the new homepage has our desired effects. There's no such inconsistency issue with the ratios that use the number of cookies in the denominator.
    
Product usage statistics like the average time the software was used in the trial period are potentially interesting features, but aren't directly related to our experiment. We might not have a strong feeling about what kind of effect the homepage will have on people that actually download the software. Stated differently, product usage isn't a direct target of the homepage manipulation. Certainly, these statistics might help us dig deeper into the reasons for observed effects after an experiment is complete. They might even point toward future changes and experiments to conduct. But in terms of experiment success, product usage shouldn't be considered an invariant or evaluation metric.
</details>

# Experiment sizing


Now that we have our main metrics selected: number of cookies as an invariant metric, and the download rate and license purchase rate (relative to number of cookies) as evaluation metrics, we should take a look at the feasibility of the experiment in terms of the amount of time it will take to run. We can use historical data as a baseline to see what it might take to detect our desired levels of change.

Recent history shows that there are about 3250 unique visitors per day, with slightly more visitors on Friday through Monday, than the rest of the week. There are about 520 software downloads per day (a .16 rate) and about 65 licenses purchased each day (a .02 rate). In an ideal case, both the download rate and license purchase rate should increase with the new homepage; a statistically significant negative change should be a sign to not deploy the homepage change. However, if only one of our metrics shows a statistically significant positive change we should be happy enough to deploy the new homepage.

### Question
Consider that we want to preserve a maximum 5% Type I error rate for falsely deploying the homepage without any actual effect. Should we apply the Bonferroni correction in this case?

<details>
    <summary>Example answer here</summary>
    
Since we're willing to deploy the homepage with an increase in only one of our two metrics (download rate, license purchase rate), we need to apply the Bonferroni correction to avoid making too many false positives due to multiple testing.
</details>

### Qustion
Let's say that we want to detect an increase of 50 downloads per day (up to 570 per day, or a .175 rate). How many days of data would we need to collect in order to get enough visitors to detect this new rate at an overall 5% Type I error rate and at 80% power?

<details>
    <summary>Hint here</summary>
    
Formula: n = ((z_null*sd_null - z_alt*sd_alt) / p_diff) ** 2
        
</details>

In [3]:
import numpy as np
import pandas as pd
import scipy.stats as stats
import matplotlib.pyplot as plt
%matplotlib inline
%config Completer.use_jedi = False

In [4]:
num_metrics = 2
alpha = 0.05
beta = 0.8

bonf_alpha = alpha / num_metrics

p_null = 0.16
p_alt = 0.175
p_diff = p_alt - p_null
print(p_diff)

0.014999999999999986


In [5]:
z_null = stats.norm.ppf(1-bonf_alpha)
z_alt = stats.norm.ppf(1-beta)
sd_null = np.sqrt(p_null*(1-p_null) + p_null*(1-p_null))
sd_alt =np.sqrt(p_null*(1-p_null) + p_alt*(1-p_alt))

In [6]:
z_null, z_alt, sd_null, sd_alt

(1.959963984540054,
 -0.8416212335729143,
 0.5184592558726288,
 0.5279914772039412)

In [7]:
n = ((z_null*sd_null - z_alt*sd_alt) / p_diff) ** 2
print(n)

9480.661238251349


So n=9481 for each group. The number of visitors is 3250 per day, which needs to be divided by 2 (control/experiment). Therefore, we need 6 days approx (9481/1625).

### Question
What if we wanted to detect an increase of 10 license purchases per day (up to 75 per day, or a .023 rate). How many days of data would we need to collect in order to get enough visitors to detect this new rate at an overall 5% Type I error rate and at 80% power?

In [8]:
alpha = 0.05 
beta = 0.8
num_metrics = 2
bonf_alpha = alpha / num_metrics

p_null = 0.02
p_alt = 0.023
p_diff = p_alt - p_null

z_null = stats.norm.ppf(1-bonf_alpha)
z_alt = stats.norm.ppf(1-beta) 
sd_null = np.sqrt(p_null*(1-p_null) + p_null*(1-p_null))
sd_alt = np.sqrt(p_null*(1-p_null) + p_alt*(1-p_alt))

In [9]:
z_null, z_alt, sd_null, sd_alt

(1.959963984540054,
 -0.8416212335729143,
 0.1979898987322333,
 0.2051121644369246)

In [10]:
n = ((z_null*sd_null - z_alt*sd_alt) / p_diff) ** 2
print(n)

34929.096074778325


Given the daily visit of 3250, we need 1625 experiment size per day for each group. This requires the collection period of 21.49 days (34929 / 1625).

<details>
    <summary>Answers for the 2 questions</summary>
    
For an overall 5% Type I error rate with Bonferroni correction and 80% power, we should require 6 days (rounded up from 5.55) to reliably detect a 50 download increase per day and 21 days (rounded up from 20.44) to detect an increase of 10 license purchases per day. 21 days is actually a convenient number since the three-week timespan helps to account for weekly cycles. In addition, the 21-day data collection period is a short enough timeframe that running the experiment is a reasonable proposition. If the required experiment length were a few weeks longer, then we might have needed to forego measuring the license purchasing rate as a critical metric.  
    
One thing that isn't accounted for in the base experiment length calculations is that there is going to be a delay between when users download the software and when they actually purchase a license. That is, when we start the experiment, there could be about seven days before a user account associated with a cookie actually comes back to make their purchase. Any purchases observed within the first week might not be attributable to either experimental condition. As a way of accounting for this, we'll run the experiment for about one week longer to allow those users who come in during the third week a chance to come back and be counted in the license purchases tally.
</details>

# Notes for Validity, Bias, and Ethics

<details>
    <summary>Expand to explore more!</summary>
        
Before getting to the data and its analysis, let's review a few of the conceptual points that go into creation of an experiment: validity, bias, and ethics.

We probably don't have too much to worry about in terms of validity. For conceptual validity, the evaluation metrics are directly aligned with the experimental goals, no abstraction needed. Internal validity is maintained by performing an experiment with properly-handled randomization and controls. We don't really need to answer to external validity since we're drawing from the full site population and there's no other population we're looking to generalize to.

As for biases, we might think of novelty bias as being a potential issue. However, we don't expect users to come back to the homepage regularly. Downloading and license purchasing are actions we expect to only occur once per user, so there's no real 'return rate' to worry about. One possibility, however, is that if more people download the software under the new homepage, the expanded user base is qualitatively different from the people who came to the page under the original homepage. This might cause more homepage hits from people looking for the support pages on the site, causing the number of unique cookies under each condition to differ. If we do see something wrong or out of place in the invariant metric (number of cookies), then this might be an area to explore in further investigations.

Finally, for ethical issues, the changes to the homepage should be benign and present no risk to users. Our experiment objectives are also clearly stated. Considering the low risks of the experiment, informed consent is at worst a minor concern; a standard popup to let visitors know that cookies are used to track user experience on the site will likely suffice. The largest ethics principle we should be concerned about is data sensitivity. We shouldn't get any sensitive data out of the cookie assignment and collection, though some information will be collected from the user when they go to download the software. No sensitive data is required for the metrics we've laid out, so what we should do is just aggregate daily visits, downloads, and purchase counts without looking at any individual outcomes.
    
</details>

# Analyze Data
Let's assume that the experiment was given the green light to go ahead, and data was collected for 29 days. As a reminder of the discussion on experiment sizing, it was found that a three-week period was needed to collect enough visitors to achieve our desired power level. Eight additional days of collection were added to allow visitors in the last week to complete their trials and come back to make a purchase – if you look at the data linked in the next paragraph, you will see that it takes about eight days before the license purchases reaches its steady level.

The data file reports the daily counts for the number of unique cookies, number of downloads, and number of license purchases attributed to each group: the experimental group with the new homepage, or the control group with the old homepage. The number of license purchases only includes purchases by users who joined after the start of the experiment, so there will be some time before the counts reach their steady state. As noted earlier, we'll assume that the potentially muddying effects of visits across multiple days, established user visits, and 'lost' cookie tracking will be ignorable, at least unless we find reason to doubt our findings.

In [11]:
DATA_PATH = 'data/homepage-experiment-data.csv'
data = pd.read_csv(DATA_PATH)
data.head()

Day  Control Cookies  Control Downloads  Control Licenses  \
0    1             1764                246                 1   
1    2             1541                234                 2   
2    3             1457                240                 1   
3    4             1587                224                 1   
4    5             1606                253                 2   

   Experiment Cookies  Experiment Downloads  Experiment Licenses  
0                1850                   339                    3  
1                1590                   281                    2  
2                1515                   274                    1  
3                1541                   284                    2  
4                1643                   292                    3

## Invariant Metric
First, we should check our invariant metric, the number of cookies assigned to each group. If there is a statistically significant difference detected, then we shouldn't move on to the evaluation metrics right away. We'd need to first dig deeper to see if there was an issue with the group-assignment procedure, or if there is something about the manipulation that affected the number of cookies observed, before we feel secure about analyzing and interpreting the evaluation metrics.

### Question
What is the p-value for the test on the number of cookies assigned to each group? Answers are at the end.

Here, we want to see how difference the number of observations for each group is when compared to null (equal count).

In [12]:
n_ctrl = data['Control Cookies'].sum()
n_expr = data['Experiment Cookies'].sum()

In [13]:
# define null
n_obs = n_ctrl + n_expr
p = 0.5

null_mean = n_obs*p
null_sd = np.sqrt(n_obs*p*(1-p))

In [14]:
# Analytic approach
z = (n_ctrl+0.5 - null_mean) / null_sd
p_val = 2 * stats.norm.cdf(z)

print(f'Z score: {z}\nP value: {p_val}')

Z score: -1.6095646049678511
P value: 0.10749294050130412


In [15]:
# Simulation approach
n_sim = 200_000
samples = np.random.binomial(n_obs, p, n_sim)

p_val = np.logical_or(samples <= n_ctrl, 
                      samples >= (n_obs-n_ctrl)).mean()

print(f'P_value: {p_val}')

P_value: 0.105835


Given the conventional alpha of 0.05 (two-tailed), the computed p value suggests not enough statistical significance to reject the null where number of observations for two groups are equal.

### Evaluation Metrics
Assuming that the invariant metric passed inspection, we can move on to the evaluation metrics: download rate and license purchasing rate. For a refresher, the download rate is the total number of downloads divided by the number of cookies, and the license purchasing rate the number of licenses divided by the number of cookies.

One tricky point to consider is that there is a seven or eight day delay between when most people download the software and when they make a purchase. There's no direct way of attributing cookies all the way through license purchases due to the daily aggregation of results, so the best we can do is to make a justified argument for handling the data. To answer the question below about the license purchasing rate, you should only take the cookies observed through day 21 as the denominator of the ratio as being responsible for all of the license purchases observed. (A more informed model of license purchasing could come up with a different handling of the data, such as including part of the day 22 cookies in the denominator.) (Note that we don't need to perform this kind of correction for the download rate, since the link between homepage visits and downloads is much closer.)



### Question
What is the p-value for the test on the download rate between groups?

In [16]:
# Number of observations
n_ctrl = data['Control Cookies'].sum()
n_expr = data['Experiment Cookies'].sum()

# Number of downloads
n_download_ctrl = data['Control Downloads'].sum() 
n_download_expr = data['Experiment Downloads'].sum()

# Proportion of downloads
p_download_ctrl = n_download_ctrl / n_ctrl
p_download_expr = n_download_expr / n_expr

print('Control group')
print(f'    Number of observation: {n_ctrl}')
print(f'    Proportion of downloads: {p_download_ctrl:.4f}')
print()
print('Experiment group')
print(f'    Number of observation: {n_expr}')
print(f'    Proportion of downloads: {p_download_expr:.4f}')

Control group
    Number of observation: 46851
    Proportion of downloads: 0.1612

Experiment group
    Number of observation: 47346
    Proportion of downloads: 0.1805


In [17]:
# Compute pooled sample proportion and standard error for null

# Pooled sample proportion
# p_total = (n_ctrl * p_download_ctrl + n_expr * p_download_expr) / (n_ctrl + n_expr)
p_download = (n_download_ctrl + n_download_expr) / (n_ctrl + n_expr)

# Pooled standard error
p_se = np.sqrt(p_download * (1-p_download) * (1/n_ctrl + 1/n_expr))

In [18]:
# Compute z score 
z = (p_download_expr-p_download_ctrl) / p_se
print(f'Z score: {z}')

# Compute p-value
pval = (1-stats.norm.cdf(z))
print(f'P value: {pval}')

Z score: 7.870833726066236
P value: 1.7763568394002505e-15


Reference at the link [here](https://stattrek.com/hypothesis-test/difference-in-proportions.aspx)

<details>
    <summary>Analyze Sample Data</summary>
      
Using sample data, complete the following computations to find the test statistic and its associated P-Value.

`Pooled sample proportion` Since the null hypothesis states that P1=P2, we use a pooled sample proportion (p) to compute the standard error of the sampling distribution.
$p$ = $(p1 * n1 + p2 * n2) / (n1 + n2)$

where p1 is the sample proportion from population 1, p2 is the sample proportion from population 2, n1 is the size of sample 1, and n2 is the size of sample 2.

`Standard error` Compute the standard error (SE) of the sampling distribution difference between two proportions.
$SE$ = $\sqrt{(p * ( 1 - p ) * ( \frac{1}{n_1} + \frac{1}{n_2} )}$

where p is the pooled sample proportion, n1 is the size of sample 1, and n2 is the size of sample 2.
    
</details>

### Question
What is the p-value for the test on the license purchasing rate between groups?

In [19]:
# Number of downloads (< Day 22)
n_21day_ctrl = data.query('Day < 22')['Control Cookies'].sum()
n_21day_expr = data.query('Day < 22')['Experiment Cookies'].sum()

# Number of licenses
n_license_ctrl = data['Control Licenses'].sum()
n_license_expr = data['Experiment Licenses'].sum()

# Proportion of licenses
p_license_ctrl = n_license_ctrl / n_21day_ctrl
p_license_expr = n_license_expr / n_21day_expr

print('Control group')
print(f'    Number of observations: {n_21day_ctrl}')
print(f'    Proportion of license purchases: {p_license_ctrl:.4f}')
print()
print('Experiment group')
print(f'    Number of downloads: {n_21day_expr}')
print(f'    Proportion of license purchases: {p_license_expr:.4f}')

Control group
    Number of observations: 33758
    Proportion of license purchases: 0.0210

Experiment group
    Number of downloads: 34338
    Proportion of license purchases: 0.0213


In [20]:
# Construct null
p_license = (n_license_ctrl + n_license_expr) / (n_21day_ctrl + n_21day_expr)
p_se_license = np.sqrt(p_license
                       * (1-p_license) 
                       * (1/n_21day_ctrl + 1/n_21day_expr))

In [21]:
# Compute z-score
z = (p_license_expr - p_license_ctrl) / p_se_license
print(f'Z score: {z}')

# Compute p-value
pval = (1-stats.norm.cdf(z)) 
print(f'P value: {pval}')

Z score: 0.2586750111658684
P value: 0.3979430008399871


<details>
    <summary>Example answers for all (3) questions here</summary>

For the test of the invariant metric, number of cookies, there were a larger number of cookies recorded in the experiment group, 47 346 vs. 46 851. This ends up generating a p-value of 0.107 (z = -1.61), which is within a reasonable range under the null hypothesis. Since we lack sufficient reason to reject the null, we can continue on to evaluating the evaluation metrics. (Note that this doesn't mean that there wasn't something actually different about the cookie counts between groups, only that we couldn't detect it if such a difference existed.)

For the first evaluation metric, download rate, there was an extremely convincing effect. An absolute increase from 0.1612 to 0.1805 results in a z-score of 7.87, well beyond any standard significance bound. However, the second evaluation metric, license purchasing rate, only shows a small increase from 0.0210 to 0.0213 (following the assumption that only the first 21 days of cookies account for all purchases). This results in a p-value of 0.398 (z = 0.26).    
</details>

# Draw Conclusion

### Review the results

Invariant metric shows the p value of 0.11, which does not prove statistic significance to reject the null where we assume no difference in control and experiment group. 

The evaluation metrics show that
- download rate is significantly different bewteen experiment and control group (one-tail, higher for experiment)
- license rate however does not show much of difference as suggested by p value of 0.398.

### Conclusion

Despite the fact that statistical significance wasn't obtained for the number of licenses purchased, the new homepage appeared to have a strong effect on the number of downloads made. Based on our goals, this seems enough to suggest replacing the old homepage with the new homepage. Establishing whether there was a significant increase in the number of license purchases, either through the rate or the increase in the number of homepage visits, will need to wait for further experiments or data collection.

One inference we might like to make is that the new homepage attracted new users who would not normally try out the program, but that these new users didn't convert to purchases at the same rate as the existing user base. This is a nice story to tell, but we can't actually say that with the data as given. In order to make this inference, we would need more detailed information about individual visitors that isn't available. However, if the software did have the capability of reporting usage statistics, that might be a way of seeing if certain profiles are more likely to purchase a license. This might then open additional ideas for improving revenue.